### DISCUSSION TITLE
Async Action performance / throughput

### DISCUSSION THREAD

**Discussion by dkulig** [hasura/graphql-engine#8714] [Question]
Hi,

I am really interested in using async actions in my product.
I'd love to understand more about the performance / throughput / scalability of the Async Actions feature.

I have a really simple use-case of my client APP calling Hasura (async action) which then delegates some business logic to the external web-hook. This webhook can finish within 1-20 seconds (some operations could be really slow, out of my control).

What is the maximum number of async actions in progress? Can I tune it w/o scaling horizontally?

Thanks!

**Comment by adas98012** [comment]
There are no specific limits on the number of async actions that can be executed simultaneously. When an async action is created, you will receive an action_id; now, you can use this action_id data using a query or subscription. So essentially, this works similar to how actions work but in async. Also, similar to how action has a default HTTP time out of 30 seconds, those will apply for async actions too. If your webhook takes more than 30 seconds, you can configure it using the custom action timeout section.

If you use async action, it is always advised to clean it up periodically. Refer to this [documentation](https://hasura.io/docs/latest/actions/logs-clean-up/) for more info on the cleanup of async actions.

**Reply by dkulig** [reply]
Thanks @adas98012 -- unfortunately this doesn't seem to be a case. I did a local test with a mocked endpoint (handler) which is delaying a response by 5 seconds and fired 100 of async actions in one go. My observation was that only 10 actions are being executed (`state = processing`) simultaneously, the other 90 actions are queued in the `created` state.

This behaviour is kinda in line with this piece of logic https://github.com/hasura/graphql-engine/blob/33b19c327c73e309b8c4e0aa31e119fe320fbc67/server/src-lib/Hasura/GraphQL/Execute/Action.hs#L717 see.
